File: ios/Classes/OrtPyannote.swift
import Flutter
import Foundation
import onnxruntime_objc
import os.log

/// OrtPyannote is a class responsible for performing speaker diarization using ONNX Runtime.
class OrtPyannote {
    /// The path to the ONNX model file.
    private let modelPath: String
    
    /// Sample rate required by Pyannote models (16kHz)
    private let sampleRate: Int64 = 16000
    
    /// Duration in samples (10 seconds window)
    private let duration: Int = 160000
    
    /// Number of speakers to detect
    private let numSpeakers: Int = 3
    
    /// Lazy instantiation of OrtSessionObjects used to manage the ONNX session.
    private lazy var model: OrtSessionObjects = {
        // Note from VAD: includeOrtExtensions is needed for Pyannote
        OrtSessionObjects(modelPath: modelPath, includeOrtExtensions: true)!
    }()
    
    /// Initializes the OrtPyannote instance with a provided model path.
    /// - Parameter modelPath: The path to the ONNX model file.
    init(modelPath: String) {
        self.modelPath = modelPath
    }
    
    /// Converts sample count to frame count.
    private func sampleToFrame(_ samples: Int) -> Int {
        return (samples - 721) / 270
    }
    
    /// Converts frame count to sample count.
    private func frameToSample(_ frames: Int) -> Int {
        return (frames * 270) + 721
    }
    
    /// Process audio data and return speaker segments
    /// - Parameter audioData: Array of audio samples as floating point values
    /// - Returns: Dictionary containing speaker segments with start/stop times
    func process(audioData: [Float]) -> [[String: Any]]? {
        os_log("Processing audio data of size: %d", audioData.count)
        do {
            guard audioData.count >= 721 else {
                os_log("Error: Audio data too short")
                return nil
            }

            let session = model.session
            let actualDuration = min(duration, audioData.count)
            let step = min(actualDuration / 2, Int(0.9 * Double(actualDuration)))
            
            var results: [[String: Any]] = []
            var isActive = Array(repeating: false, count: numSpeakers)
            var startSamples = Array(repeating: 0, count: numSpeakers)
            var currentSamples = 721
            
            // Calculate overlap size
            let overlap = sampleToFrame(duration - step)
            var overlapChunk = Array(repeating: Array(repeating: 0.0, count: numSpeakers), count: overlap)
            
            // Create sliding windows
            let windows = createSlidingWindows(audioData: audioData, windowSize: actualDuration, stepSize: step)
            
            for (windowIndex, (windowSize, window)) in windows.enumerated() {
                os_log("Processing window %d with size %d", windowIndex, windowSize)
                
                // Following VAD pattern for 3D tensor creation
                let windowData = NSMutableData(
                    bytes: window,
                    length: window.count * MemoryLayout<Float>.size
                )
                
                // Create input tensor following VAD's 3D pattern
                let inputTensor = try ORTValue(
                    tensorData: windowData,
                    elementType: .float,
                    shape: [NSNumber(value: 1), NSNumber(value: 1), NSNumber(value: window.count)]
                )
                
                let inputs = ["input_values": inputTensor]
                let outputNames = Set(["logits"])
                
                // Run inference following VAD pattern
                let outputs = try session.run(
                    withInputs: inputs,
                    outputNames: outputNames,
                    runOptions: nil
                )
                
                // Extract outputs following VAD pattern
                guard let outputTensor = outputs["logits"],
                      let outputData = try? outputTensor.tensorData() as Data else {
                    continue
                }
                
                // Process output frames
                let frameOutputs = processOutputData(outputData)
                var processedFrames = frameOutputs
                
                // Handle frame overlapping
                if windowIndex > 0 {
                    processedFrames = reorderAndBlend(overlapChunk: overlapChunk, newFrames: frameOutputs)
                }
                
                if windowIndex < windows.count - 1 {
                    overlapChunk = Array(processedFrames.suffix(overlap))
                    processedFrames = Array(processedFrames.dropLast(overlap))
                } else {
                    processedFrames = Array(processedFrames.prefix(min(processedFrames.count, sampleToFrame(windowSize))))
                }
                
                // Track speaker segments
                for probs in processedFrames {
                    currentSamples += 270
                    for speaker in 0..<numSpeakers {
                        if isActive[speaker] {
                            if probs[speaker] < 0.5 {
                                results.append([
                                    "speaker": speaker,
                                    "start": Double(startSamples[speaker]) / Double(sampleRate),
                                    "stop": Double(currentSamples) / Double(sampleRate)
                                ])
                                isActive[speaker] = false
                            }
                        } else if probs[speaker] > 0.5 {
                            startSamples[speaker] = currentSamples
                            isActive[speaker] = true
                        }
                    }
                }
            }
            
            // Handle any remaining active speakers
            for speaker in 0..<numSpeakers {
                if isActive[speaker] {
                    results.append([
                        "speaker": speaker,
                        "start": Double(startSamples[speaker]) / Double(sampleRate),
                        "stop": Double(currentSamples) / Double(sampleRate)
                    ])
                }
            }
            
            return results
            
        } catch {
            os_log("Error in process: %{public}s", error.localizedDescription)
            for symbol in Thread.callStackSymbols {
                os_log("Stack trace: %{public}s", symbol)
            }
            return nil
        }
    }
    
    private func createSlidingWindows(audioData: [Float], windowSize: Int, stepSize: Int) -> [(Int, [Float])] {
        var windows: [(Int, [Float])] = []
        var start = 0
        
        while start <= audioData.count - windowSize {
            let window = Array(audioData[start..<(start + windowSize)])
            windows.append((windowSize, window))
            start += stepSize
        }
        
        if audioData.count < windowSize || (audioData.count - windowSize) % stepSize > 0 {
            var lastWindow = Array(audioData[start...])
            let lastWindowSize = lastWindow.count
            
            if lastWindow.count < windowSize {
                lastWindow.append(contentsOf: Array(repeating: 0.0, count: windowSize - lastWindow.count))
            }
            
            windows.append((lastWindowSize, lastWindow))
        }
        
        return windows
    }
    
    private func processOutputData(_ outputData: Data) -> [[Double]] {
        let floats = outputData.toArray(type: Float.self)
        var frameOutputs: [[Double]] = []
        
        let numCompleteFrames = floats.count / 7
        for frame in 0..<numCompleteFrames {
            let i = frame * 7
            let probs = Array(floats[i..<(i + 7)]).map { exp(Double($0)) }
            
            var speakerProbs = Array(repeating: 0.0, count: numSpeakers)
            speakerProbs[0] = probs[1] + probs[4] + probs[5] // spk1
            speakerProbs[1] = probs[2] + probs[4] + probs[6] // spk2
            speakerProbs[2] = probs[3] + probs[5] + probs[6] // spk3
            
            frameOutputs.append(speakerProbs)
        }
        
        return frameOutputs
    }
    
    private func reorderAndBlend(overlapChunk: [[Double]], newFrames: [[Double]]) -> [[Double]] {
        var reorderedFrames = reorder(x: overlapChunk, y: newFrames)
        
        // Blend overlapping sections
        for i in 0..<overlapChunk.count {
            for j in 0..<numSpeakers {
                reorderedFrames[i][j] = (reorderedFrames[i][j] + overlapChunk[i][j]) / 2.0
            }
        }
        
        return reorderedFrames
    }
    
    private func reorder(x: [[Double]], y: [[Double]]) -> [[Double]] {
        let perms = generatePermutations(n: numSpeakers)
        let yTransposed = transpose(y)
        
        var minDiff = Double.infinity
        var bestPerm = y
        
        for perm in perms {
            var permuted = Array(repeating: Array(repeating: 0.0, count: numSpeakers), count: y.count)
            for i in 0..<y.count {
                for j in 0..<numSpeakers {
                    permuted[i][j] = yTransposed[perm[j]][i]
                }
            }
            
            var diff = 0.0
            for i in 0..<x.count {
                for j in 0..<numSpeakers {
                    diff += abs(x[i][j] - permuted[i][j])
                }
            }
            
            if diff < minDiff {
                minDiff = diff
                bestPerm = permuted
            }
        }
        
        return bestPerm
    }
    
    private func generatePermutations(n: Int) -> [[Int]] {
        if n == 1 {
            return [[0]]
        }
        
        var result: [[Int]] = []
        for i in 0..<n {
            let subPerms = generatePermutations(n: n - 1)
            for var perm in subPerms {
                perm = [i] + perm.map { $0 >= i ? $0 + 1 : $0 }
                result.append(perm)
            }
        }
        return result
    }
    
    private func transpose(_ matrix: [[Double]]) -> [[Double]] {
        guard !matrix.isEmpty else { return [] }
        let rows = matrix.count
        let cols = matrix[0].count
        var result = Array(repeating: Array(repeating: 0.0, count: rows), count: cols)
        
        for i in 0..<rows {
            for j in 0..<cols {
                result[j][i] = matrix[i][j]
            }
        }
        
        return result
    }
}

File: android/src/main/kotlin/com/telosnex/fonnx/OrtPyannote.kt
package com.telosnex.fonnx

import ai.onnxruntime.*
import android.util.Log
import kotlin.math.exp
import kotlin.math.min

class OrtPyannote(private val modelPath: String) {
    private val model: OrtSessionObjects by lazy {
        OrtSessionObjects(modelPath, true)
    }

    companion object {
        private const val SAMPLE_RATE = 16000
        private const val DURATION = 160000 // 10 seconds window
        private const val NUM_SPEAKERS = 3
    }

    fun process(audioData: FloatArray): List<Map<String, Any>>? {
        try {
            val ortEnv = model.ortEnv
            val session = model.ortSession
            val step = min(DURATION / 2, (0.9 * DURATION).toInt())
            val results = mutableListOf<Map<String, Any>>()
            
            val isActive = BooleanArray(NUM_SPEAKERS) { false }
            val startSamples = IntArray(NUM_SPEAKERS) { 0 }
            var currentSamples = 721

            // Calculate overlap size
            val overlap = sampleToFrame(DURATION - step)
            var overlapChunk = Array(overlap) { DoubleArray(NUM_SPEAKERS) { 0.0 } }

            // Create sliding windows
            val windows = createSlidingWindows(audioData, DURATION, step)

            session?.let {
                for ((windowIndex, window) in windows.withIndex()) {
                    val (windowSize, windowData) = window
                    // Log.d("[OrtPyannote.kt]", "Processing window $windowIndex of ${windows.size}; this window has ${windowData.size} samples")
                    // Prepare input tensor
                    val inputs = mutableMapOf<String, OnnxTensor>()
                    inputs["input_values"] = OnnxTensor.createTensor(
                        ortEnv,
                        arrayOf(arrayOf(windowData))
                    )

                    val startTime = System.currentTimeMillis()
                    val rawResult = it.run(inputs, setOf("logits"))
                    val endTime = System.currentTimeMillis()
                    // Log.d("[OrtPyannote.kt]", "Inference time: ${endTime - startTime} ms")

                    // Process output
                    val logits = rawResult.get("logits").get().value as Array<Array<FloatArray>>
                    // Log.d("[OrtPyannote.kt]", "Logits shape: ${logits.size} x ${logits[0].size} x ${logits[0][0].size}")
                    // Log.d("[OrtPyannote.kt]", "Logits: ${logits[0][0].joinToString(", ")}")
                    var frameOutputs = processOutputData(logits)
                    // Log.d("[OrtPyannote.kt]", "Frame outputs: ${frameOutputs.size}")

                    // Handle overlapping
                    if (windowIndex > 0) {
                        frameOutputs = reorderAndBlend(overlapChunk, frameOutputs)
                    }

                    if (windowIndex < windows.size - 1) {
                        overlapChunk = frameOutputs.takeLast(overlap).toTypedArray()
                        frameOutputs = frameOutputs.dropLast(overlap)
                    } else {
                        frameOutputs = frameOutputs.take(
                            min(frameOutputs.size, sampleToFrame(windowSize))
                        )
                    }

                    // Track speaker segments
                    for (probs in frameOutputs) {
                        currentSamples += 270
                        for (speaker in 0 until NUM_SPEAKERS) {
                            if (isActive[speaker]) {
                                if (probs[speaker] < 0.5) {
                                    results.add(mapOf(
                                        "speaker" to speaker,
                                        "start" to (startSamples[speaker].toDouble() / SAMPLE_RATE),
                                        "stop" to (currentSamples.toDouble() / SAMPLE_RATE)
                                    ))
                                    isActive[speaker] = false
                                }
                            } else if (probs[speaker] > 0.5) {
                                startSamples[speaker] = currentSamples
                                isActive[speaker] = true
                            }
                        }
                    }
                }
            }

            // Handle any remaining active speakers
            for (speaker in 0 until NUM_SPEAKERS) {
                if (isActive[speaker]) {
                    results.add(mapOf(
                        "speaker" to speaker,
                        "start" to (startSamples[speaker].toDouble() / SAMPLE_RATE),
                        "stop" to (currentSamples.toDouble() / SAMPLE_RATE)
                    ))
                }
            }

            return results
        } catch (e: Exception) {
            Log.e("[OrtPyannote.kt]", "Error in process: ${e.message}")
            e.printStackTrace()
            return null
        }
    }

    private fun sampleToFrame(samples: Int): Int {
        return (samples - 721) / 270
    }

    private fun frameToSample(frames: Int): Int {
        return (frames * 270) + 721
    }

    private fun createSlidingWindows(
        audioData: FloatArray,
        windowSize: Int,
        stepSize: Int
    ): List<Pair<Int, FloatArray>> {
        val windows = mutableListOf<Pair<Int, FloatArray>>()
        var start = 0

        while (start <= audioData.size - windowSize) {
            windows.add(Pair(
                windowSize,
                audioData.slice(start until start + windowSize).toFloatArray()
            ))
            start += stepSize
        }

        // Handle last window if needed
        if (audioData.size < windowSize || (audioData.size - windowSize) % stepSize > 0) {
            val lastWindow = audioData.slice(start until audioData.size).toFloatArray()
            val lastWindowSize = lastWindow.size

            if (lastWindow.size < windowSize) {
                val paddedWindow = FloatArray(windowSize)
                lastWindow.copyInto(paddedWindow)
                windows.add(Pair(lastWindowSize, paddedWindow))
            } else {
                windows.add(Pair(lastWindowSize, lastWindow))
            }
        }

        return windows
    }

    private fun processOutputData(logits: Array<Array<FloatArray>>): List<DoubleArray> {
        val frameOutputs = mutableListOf<DoubleArray>()
        val batchLogits = logits[0]  // First batch
        
        // Process each frame (589 frames)
        for (frame in batchLogits.indices) {
            val frameLogits = batchLogits[frame]  // Get the 7 logits for this frame
            val probs = frameLogits.map { exp(it.toDouble()) }
            
            val speakerProbs = DoubleArray(NUM_SPEAKERS)
            // Combine probabilities for each speaker
            // spk1: solo (1) + with_spk2 (4) + with_spk3 (5)
            speakerProbs[0] = probs[1] + probs[4] + probs[5]
            // spk2: solo (2) + with_spk1 (4) + with_spk3 (6)
            speakerProbs[1] = probs[2] + probs[4] + probs[6]
            // spk3: solo (3) + with_spk1 (5) + with_spk2 (6)
            speakerProbs[2] = probs[3] + probs[5] + probs[6]
            
            frameOutputs.add(speakerProbs)
        }
    
        return frameOutputs
    }

    private fun reorderAndBlend(
        overlapChunk: Array<DoubleArray>,
        newFrames: List<DoubleArray>
    ): List<DoubleArray> {
        val reorderedFrames = reorder(overlapChunk.toList(), newFrames)
        
        // Blend overlapping sections
        for (i in overlapChunk.indices) {
            for (j in 0 until NUM_SPEAKERS) {
                reorderedFrames[i][j] = (reorderedFrames[i][j] + overlapChunk[i][j]) / 2.0
            }
        }
        
        return reorderedFrames
    }

    private fun reorder(
        x: List<DoubleArray>,
        y: List<DoubleArray>
    ): List<DoubleArray> {
        val perms = generatePermutations(NUM_SPEAKERS)
        val yTransposed = transpose(y)
        
        var minDiff = Double.POSITIVE_INFINITY
        var bestPerm = y
        
        for (perm in perms) {
            val permuted = List(y.size) { i ->
                DoubleArray(NUM_SPEAKERS) { j ->
                    yTransposed[perm[j]][i]
                }
            }
            
            var diff = 0.0
            for (i in x.indices) {
                for (j in 0 until NUM_SPEAKERS) {
                    diff += kotlin.math.abs(x[i][j] - permuted[i][j])
                }
            }
            
            if (diff < minDiff) {
                minDiff = diff
                bestPerm = permuted
            }
        }
        
        return bestPerm
    }

    private fun generatePermutations(n: Int): List<List<Int>> {
        if (n == 1) {
            return listOf(listOf(0))
        }
        
        val result = mutableListOf<List<Int>>()
        for (i in 0 until n) {
            val subPerms = generatePermutations(n - 1)
            for (perm in subPerms) {
                val newPerm = listOf(i) + perm.map { if (it >= i) it + 1 else it }
                result.add(newPerm)
            }
        }
        return result
    }

    private fun transpose(matrix: List<DoubleArray>): List<DoubleArray> {
        if (matrix.isEmpty()) return emptyList()
        val rows = matrix.size
        val cols = matrix[0].size
        
        return List(cols) { j ->
            DoubleArray(rows) { i ->
                matrix[i][j]
            }
        }
    }
}